persist,bench: add end_to_end benchmark

aljoscha opened the pull request:
This only benchmarks the part of persisted_source() that replays an existing snapshot, the live-listen part is not exercised at all.

The added `upper()` method returns an owned `Antichain`. I could return an `AntichainRef`, but then I'd have to construct an owned antichain when constructing the snapshot because I can't construct one on the fly in `upper()`. Alternatively, I could add a `read_upper(&mut Antichain<_>)` that reads into a provided `Antichain`.

I've also changed `persisted_source()` to not create the listen-source when taking the snapshot fails. I had to move (and duplicate) some code around for that and it makes stuff slightly more complicated. If we agree that it's the right approach I can try and clean that up a bit. (this is in a separate `fixup!` commit, so you can clearly see what I mean)

Closes #7878

aljoscha commented:
For kicks, I ran the benchmark with 100 mb data size. I noticed that the write overhead gets quite big:
 - for the 4b-keys bench, the blob directory grows to 5GB
 - for the 4kb-keys bench, it grows to 310MB

Is that expected? It could also be that something in my write logic is wrong and I'm writing more data than I think I'm writing. 

Also: writing the many small records takes quite a while.

ruchirK commented on src/persist/benches/end_to_end.rs:
```diff
@@ -196,24 +200,39 @@ fn write_test_data(
-        .collect();
+    // write data in batches
+
+    let step_size = 1000;
```
same as before, now some key sizes will have many more batches than others. that + the fact that we don't physically delete batches i think is also part of why the 4b test uses more disk (the more batches you have the more compactions you will do which without physical deletions just makes more batches)

danhhz replied:
yeah, might be interesting to chunk based on the size of data instead of number of records

ruchirK reviewed (commented):
I didn't really review the operator changes in much depth, and stuck instead to the upper and test changes. Things look good so far. There's some stray printlns and debug statements that I didn't comment on individually.

The most interesting bit for me is how the test changes as key sizes change. Left some comments about that.

danhhz reviewed (commented):
This is great! I only have small comments.

Also, please include the output of `cargo bench` on your laptop for these in a commit message somewhere. It's nice to have rough ideas of where benchmarks are at, even if it's not going to compare perfectly (different laptops, sets of background traffic, etc)

> I've also changed persisted_source() to not create the listen-source when taking the snapshot fails. I had to move (and duplicate) some code around for that and it makes stuff slightly more complicated. If we agree that it's the right approach I can try and clean that up a bit. (this is in a separate fixup! commit, so you can clearly see what I mean)

The operator stuff seems fine to me as-is, that whole section of code is messy right now and probably needs a different approach to be any sort of reasonable

> For kicks, I ran the benchmark with 100 mb data size. I noticed that the write overhead gets quite big:
> 
> for the 4b-keys bench, the blob directory grows to 5GB
> for the 4kb-keys bench, it grows to 310MB
> Is that expected? It could also be that something in my write logic is wrong and I'm writing more data than I think I'm writing.

A few of factors here. One is the +8 + 8 thing Ruchir caught. Another is that you probably caught a window where we'd hooked up compactions, but hadn't started deleting the data that got compacted. If you pull past 7964, this should get better. Finally, our write and space amplification are hot garbage right now. This should get much better in the coming weeks.

> Also: writing the many small records takes quite a while.

There's for sure some low hanging fruit here. No one has really looked at it too closely yet since we've been focused so far on eliminating the obvious bottlenecks. We only recently stopped fsync'ing between each write.

> The added upper() method returns an owned Antichain. I could return an AntichainRef, but then I'd have to construct an owned antichain when constructing the snapshot because I can't construct one on the fly in upper(). Alternatively, I could add a read_upper(&mut Antichain<_>) that reads into a provided Antichain.

It's only called once at dataflow construction time, so a clone seems totally fine

aljoscha commented on src/persist/benches/end_to_end.rs:
```diff
@@ -0,0 +1,244 @@
+    rand::thread_rng()
+        .sample_iter(&Alphanumeric)
+        .take(len.try_into().unwrap())
+        .map(|c| c as u8)
```
from u8, I copied this `generator/bytes.rs` without changing. :see_no_evil:

danhhz replied:
Huh, I'm surprised clippy doesn't catch that. Maybe we have whatever lint it would be turned off

danhhz commented on src/persist/src/operators/source.rs:
```diff
@@ -59,72 +61,131 @@ where
+        let worker_index = scope.index();
+        let activator = scope.activator_for(&info.address[..]);
+
+        // For Antichain<u64> there can only ever be one element, because
```
Fascinating, I think you're right. My mental model was that Capability was/has an Antichain, but looks like that's wrong. What you have here is good for now

aljoscha commented on src/persist/src/indexed/mod.rs:
```diff
@@ -1042,6 +1042,16 @@ impl IndexedSnapshot {
+        //
+        // After writing some data an sealing, UnsealedSnaphot.upper() will be at [], meaning empty
+        // frontier, meaning: we're at +Inf. Which is not what we want.
+        self.1.upper()
```
@ruchirK & @danhhz: This is still unclear. I think it's correct two take the upper of the trace, but I'm not 100% sure. On a reader that is just opened for listening, I imagine the unsealed to always be empty and the trace is what matters. Whenever we seal the unsealeds, the listener would also get the `Seal` message.

danhhz replied:
So it turned out I led you astray in my previous review, my apologies! The ts_upper on UnsealedSnapshot is an optional filter used by one internal codepath. We should probably rename that to ts_upper_filter to avoid the same confusion in the future. (ts_lower is also a filter, but this one is always used. if we rename ts_upper, we should probably rename ts_lower as well to match.)

What we actually want here is to explicitly put the sealed timestamp as a field on IndexedSnapshot like we already do for the SeqNo. You can get it in `do_snapshot` by calling `get_seal` on the `Trace` that's already available in that method

aljoscha commented on src/persist/src/operators/source.rs:
```diff
@@ -59,72 +61,131 @@ where
+        // that, though, I don't think we can downgrade this capability because
+        // downgrading can only happen when `old_ts` <= `new_ts`, which is not
+        // the case when an Antichain has multiple elements.
+        for element in lower_filter.iter() {
```
@ruchirK & @danhhz A tricky part.

aljoscha commented:
I rebased this, incorporated the feedback and force pushed. I hope that's alright. If you prefer I not force-push in the future and instead only add commits on top I can also do that. :ok_hand: 

I tagged you on two inline things that seem contentious.

The one other thing I'm still thinking about is removing the `lower_filter` part where I filter records in the listen source. I now saw https://github.com/MaterializeInc/materialize/blob/c1b33f831ca92ab12d44cb164ae5d1bfa87f90fc/src/persist/src/operators/source.rs#L299, and that was my naive attempt to fix it, without knowing of this test. I think using the seq-no is the right approach, though. So I would remove the filtering. WDYT?

Also: thanks a lot, the comments were very helpful!

danhhz reviewed (approved):
LGTM modulo the seal ts fix

> If you prefer I not force-push in the future and instead only add commits on top I can also do that. 👌

For anything moderately complex, I'd love append-only commits. I tend to look fairly carefully at the changes made in response to review comments and that makes it quite a bit easier with the github CR tooling.

Also apparently it ate some of my written-but-not-yet-sent comments when you force pushed :(

> The one other thing I'm still thinking about is removing the `lower_filter` part where I filter records in the listen source. I now saw https://github.com/MaterializeInc/materialize/blob/c1b33f831ca92ab12d44cb164ae5d1bfa87f90fc/src/persist/src/operators/source.rs#L299, and that was my naive attempt to fix it, without knowing of this test. I think using the seq-no is the right approach, though. So I would remove the filtering. WDYT?

Either way. SeqNo is definitely the way to go here, but what you have is better than what we had before.

aljoscha commented on src/persist/src/operators/source.rs:
```diff
@@ -149,21 +210,92 @@ mod tests {
+        };
+
+        let guards = timely::execute(Config::thread(), move |worker| {
+            let p = create_runtime(temp_dir.path()).expect("missing runtime");
```
I tried that but gave up because I have to create the runtime inside the timely thread and I couldn't move the registry into the thread. I have to create the runtime inside timely because it needs to be dropped with timely, otherwise the thread wouldn't stop. It's the same problem we talked about a while ago.

I solved it now by wrapping the registry in an `Arc<Mutex<>>`, and handing it to the timely thread.

aljoscha replied:
This is also a response to the earlier comment about using `MemRegistry::open`

danhhz replied:
Oh! You can also just call `stop` on the RuntimeClient, that will do the same thing as dropping them all. It should also allow you to call MemRegistry::open again with the same path for the read part

aljoscha commented on src/persist/src/operators/source.rs:
```diff
@@ -149,21 +210,92 @@ mod tests {
+            6
+        };
+
+        let guards = timely::execute(Config::thread(), move |worker| {
```
You're right! I thought I needed `execute()` because I need manuall control over stepping, but it also works with `execute_directly()` because returning from the closure will shut down the sources, meaning the `step()` inside `execute_directly()` will also finish.

aljoscha commented:
Thanks! I'll squash and push again, please ignore any further changes because it's just re-arranging/squashing of commits.
